File: maxsim/bench.rs
//! Benchmark for MaxSim (ColBERT-style late-interaction) operations
//!
//! MaxSim computes: score = Sigma_i max_j dot(q_i, d_j)
//! Used in ColBERT and other late-interaction retrieval models.
//!
//! Competitors:
//! - numkong (MaxSimPackedMatrix::score on pre-packed tensors)
//! - ndarray (matmul-based: Q @ D.T then row-wise max + sum)
//!
//! Run with:
//! ```bash
//! cargo bench --features bench_maxsim --bench bench_maxsim
//! NUMWARS_FILTER="f16|bf16|f32" cargo bench --features bench_maxsim
//! ```
//!
//! Environment variables:
//! - NUMWARS_DIMS_DEPTH: Shared dimension k (default: 2048)
//! - NUMWARS_DIMS_HEIGHT: Query count m (default: 2048)
//! - NUMWARS_DIMS_WIDTH: Document count n (default: 2048)
//! - NUMWARS_FILTER: Regex to filter benchmark names
//!
//! Benchmark naming: maxsim/{dtype}
//! Examples: maxsim/f32, maxsim/bf16, maxsim/f16

#[path = "../utils.rs"]
mod utils;

use criterion::measurement::WallTime;
use criterion::{criterion_group, criterion_main, BenchmarkGroup, Criterion, Throughput};
use ndarray::Array2;
use numkong::prelude::*;
use numkong::{bf16, capabilities, f16, MaxSim, MaxSimPackedMatrix};
use std::hint::black_box;
use utils::*;

// region: Per-Library Run Traits

trait RunNumKong: Sized {
    fn run(
        _group: &mut BenchmarkGroup<'_, WallTime>,
        _query_vectors: &[Self],
        _document_vectors: &[Self],
        _qc: usize,
        _dc: usize,
        _dim: usize,
    ) {
    }
}

fn run_numkong<T: MaxSim + Clone>(
    group: &mut BenchmarkGroup<'_, WallTime>,
    query_vectors: &[T],
    document_vectors: &[T],
    qc: usize,
    dc: usize,
    dim: usize,
) {
    let query_tensor = Tensor::<T>::try_from_slice(query_vectors, &[qc, dim]).expect("Failed to create query tensor");
    let document_tensor =
        Tensor::<T>::try_from_slice(document_vectors, &[dc, dim]).expect("Failed to create document tensor");
    let packed_queries = MaxSimPackedMatrix::<T>::try_pack(&query_tensor.view()).expect("Failed to pack queries");
    let packed_documents =
        MaxSimPackedMatrix::<T>::try_pack(&document_tensor.view()).expect("Failed to pack documents");
    group.bench_function("numkong", |bench| {
        bench.iter(|| black_box(packed_queries.try_score(&packed_documents).expect("scoring failed")))
    });
}

impl<T: MaxSim + Clone> RunNumKong for T {
    fn run(
        group: &mut BenchmarkGroup<'_, WallTime>,
        query_vectors: &[Self],
        document_vectors: &[Self],
        qc: usize,
        dc: usize,
        dim: usize,
    ) {
        run_numkong(group, query_vectors, document_vectors, qc, dc, dim);
    }
}

trait RunNdarray: Sized {
    fn run(
        _group: &mut BenchmarkGroup<'_, WallTime>,
        _query_vectors: &[Self],
        _document_vectors: &[Self],
        _qc: usize,
        _dc: usize,
        _dim: usize,
    ) {
    }
}

impl RunNdarray for f32 {
    fn run(
        group: &mut BenchmarkGroup<'_, WallTime>,
        query_vectors: &[f32],
        document_vectors: &[f32],
        qc: usize,
        dc: usize,
        dim: usize,
    ) {
        let query_matrix =
            Array2::from_shape_vec((qc, dim), query_vectors.to_vec()).expect("Failed to build query matrix");
        let document_matrix =
            Array2::from_shape_vec((dc, dim), document_vectors.to_vec()).expect("Failed to build document matrix");
        group.bench_function("ndarray", |bench| {
            bench.iter(|| {
                let scores_matrix = query_matrix.dot(&document_matrix.t());
                let total: f32 = scores_matrix
                    .rows()
                    .into_iter()
                    .map(|row| row.iter().copied().fold(f32::NEG_INFINITY, f32::max))
                    .sum();
                black_box(total)
            })
        });
    }
}

impl RunNdarray for f16 {}
impl RunNdarray for bf16 {}

// endregion

// region: Generic Helpers

fn bench_maxsim_dtype<T>(
    c: &mut Criterion,
    dtype: &str,
    query_count: usize,
    document_count: usize,
    dimension: usize,
    init: T,
) where
    T: Clone + RunNumKong + RunNdarray + 'static,
{
    let name = format!("maxsim/{dtype}");
    if !should_run_benchmark(&name) {
        return;
    }

    let mut group = c.benchmark_group(name);
    let query_vectors = vec![init.clone(); query_count * dimension];
    let document_vectors = vec![init; document_count * dimension];
    group.throughput(Throughput::Bytes(
        ((query_count + document_count) * dimension * std::mem::size_of::<T>()) as u64,
    ));

    <T as RunNumKong>::run(
        &mut group,
        &query_vectors,
        &document_vectors,
        query_count,
        document_count,
        dimension,
    );
    <T as RunNdarray>::run(
        &mut group,
        &query_vectors,
        &document_vectors,
        query_count,
        document_count,
        dimension,
    );

    group.finish();
}

// endregion

// region: Benchmarks

/// Benchmark MaxSim scoring.
pub fn bench_maxsim(c: &mut Criterion) {
    capabilities::configure_thread();
    let dimension = get_matrix_dims_depth();
    let query_count = get_matrix_dims_height();
    let document_count = get_matrix_dims_width();

    bench_maxsim_dtype(c, "f32", query_count, document_count, dimension, 1.0f32);
    bench_maxsim_dtype(c, "bf16", query_count, document_count, dimension, bf16::from_f32(1.0));
    bench_maxsim_dtype(c, "f16", query_count, document_count, dimension, f16::from_f32(1.0));
}

// endregion

// region: Main

criterion_group! {
    name = benches;
    config = utils::configure_criterion();
    targets = bench_maxsim
}
criterion_main!(benches);

// endregion

File: maxsim/bench.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
MaxSim (ColBERT-style late-interaction) benchmarks: NumKong vs NumPy.

Computes score = Sigma_i max_j dot(q_i, d_j) using NumKong's maxsim API
and a NumPy matmul + row-max + sum baseline.

Can be run with uv:
    uv run --with numkong,numpy,tabulate,ml_dtypes maxsim/bench.py

Or with traditional pip:
    pip install -e ".[maxsim]"
    python maxsim/bench.py

Environment variables:
    NUMWARS_FILTER - Regex filter for benchmark names
    NUMWARS_DIMS_WIDTH - Document count (default: 2048)
    NUMWARS_DIMS_HEIGHT - Query count (default: 2048)
    NUMWARS_DIMS_DEPTH - Shared dimension (default: 2048)
"""

import argparse
import json
import os
import re
import sys
from dataclasses import dataclass
from typing import List

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Import utils first — it sets single-threaded env vars (OMP_NUM_THREADS, etc.)
# that must be in place before numpy/OpenBLAS initializes.
try:
    from utils import (
        add_common_args,
        calculate_gso_per_sec,
        format_duration,
        get_matrix_dims_depth,
        get_matrix_dims_height,
        get_matrix_dims_width,
        measure_average_duration,
        numkong_dtype_name,
        parse_numpy_dtype,
        print_results_table,
        should_run_benchmark,
    )
except ImportError:
    print("Error: Could not import utils.py. Make sure it's in the parent directory.")
    sys.exit(1)

import numpy as np
import tabulate

try:
    import numkong as nk
except ImportError:
    print("Error: numkong not found. Install with: pip install numkong")
    sys.exit(1)


@dataclass
class BenchmarkResult:
    library: str
    input_dtype: str
    output_dtype: str
    display_signature: str
    height: int
    width: int
    depth: int
    duration_secs: float
    gso_per_sec: float


ACCUMULATOR_DTYPE = {
    "f16": "f32",
    "bf16": "f32",
    "f32": "f64",
}


def display_signature_name(input_dtype: str, output_dtype: str) -> str:
    return f"{input_dtype} \u2192 {output_dtype}"


def build_matrix(shape: tuple[int, int], dtype_str: str, seed: int) -> np.ndarray:
    rng = np.random.default_rng(seed)
    dtype = parse_numpy_dtype(dtype_str)
    data = rng.uniform(-1.0, 1.0, size=shape).astype(np.float32)
    return data.astype(dtype)


def benchmark_numkong(
    height: int, width: int, depth: int, dtype_str: str, warmup: float, profile: float
) -> BenchmarkResult:
    queries = build_matrix((height, depth), dtype_str, 42)
    documents = build_matrix((width, depth), dtype_str, 43)
    nk_dtype = numkong_dtype_name(dtype_str)
    packed_queries = nk.maxsim_pack(queries, dtype=nk_dtype)
    packed_documents = nk.maxsim_pack(documents, dtype=nk_dtype)
    output_dtype = ACCUMULATOR_DTYPE.get(dtype_str, dtype_str)

    duration = measure_average_duration(
        lambda: nk.maxsim_packed(packed_queries, packed_documents),
        warmup,
        profile,
    )
    num_operations = 2 * height * width * depth
    return BenchmarkResult(
        library="NumKong",
        input_dtype=dtype_str,
        output_dtype=output_dtype,
        display_signature=display_signature_name(dtype_str, output_dtype),
        height=height,
        width=width,
        depth=depth,
        duration_secs=duration,
        gso_per_sec=calculate_gso_per_sec(num_operations, duration),
    )


def benchmark_numpy(
    height: int, width: int, depth: int, warmup: float, profile: float
) -> BenchmarkResult:
    queries = build_matrix((height, depth), "f32", 42)
    documents = build_matrix((width, depth), "f32", 43)
    out = np.empty((height, width), dtype=np.float32)

    def _numpy_maxsim():
        np.matmul(queries, documents.T, out=out)
        return out.max(axis=1).sum()

    duration = measure_average_duration(_numpy_maxsim, warmup, profile)
    num_operations = 2 * height * width * depth
    return BenchmarkResult(
        library="NumPy",
        input_dtype="f32",
        output_dtype="f32",
        display_signature=display_signature_name("f32", "f32"),
        height=height,
        width=width,
        depth=depth,
        duration_secs=duration,
        gso_per_sec=calculate_gso_per_sec(num_operations, duration),
    )


def result_to_entry(result: BenchmarkResult) -> dict:
    return {
        "suite": "maxsim",
        "workload": "maxsim",
        "benchmark_id": f"maxsim/{result.library.lower()}/{result.input_dtype}/{result.height}x{result.width}x{result.depth}",
        "library": result.library,
        "dtype": result.input_dtype,
        "input_dtype": result.input_dtype,
        "output_dtype": result.output_dtype,
        "display_dtype": result.display_signature,
        "display_signature": result.display_signature,
        "accumulator_dtype": result.output_dtype,
        "height": result.height,
        "width": result.width,
        "depth": result.depth,
        "primary_value": result.gso_per_sec,
        "unit": "GSO/s",
        "duration_secs": result.duration_secs,
    }


def main():
    parser = argparse.ArgumentParser(
        description="Benchmark MaxSim (ColBERT-style late-interaction) scoring",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    add_common_args(parser)
    parser.add_argument(
        "--output-format",
        choices=["table", "json"],
        default="table",
        help="Choose human-readable table output or machine-readable JSON (default: table).",
    )
    args = parser.parse_args()

    height = get_matrix_dims_height()
    width = get_matrix_dims_width()
    depth = get_matrix_dims_depth()

    filter_pattern = None
    if args.filter:
        try:
            filter_pattern = re.compile(args.filter)
        except re.error as e:
            print(f"Warning: Invalid regex pattern '{args.filter}': {e}")

    metadata = {
        "height": height,
        "width": width,
        "depth": depth,
        "numkong_version": getattr(nk, "__version__", None),
        "numpy_version": np.__version__,
    }

    candidates = [
        ("numkong", "f16"),
        ("numkong", "bf16"),
        ("numkong", "f32"),
        ("numpy", "f32"),
    ]

    all_results: List[BenchmarkResult] = []
    for library_slug, dtype in candidates:
        benchmark_name = f"maxsim/{library_slug}/{dtype}/{height}x{width}x{depth}"
        if not should_run_benchmark(benchmark_name, filter_pattern):
            continue
        if args.output_format == "table":
            print(f"Benchmarking {benchmark_name}")
        if library_slug == "numkong":
            all_results.append(
                benchmark_numkong(
                    height, width, depth, dtype, args.warmup, args.time_limit
                )
            )
        else:
            all_results.append(
                benchmark_numpy(height, width, depth, args.warmup, args.time_limit)
            )

    if args.output_format == "json":
        print(
            json.dumps(
                {
                    "suite": "maxsim",
                    "settings": metadata,
                    "results": [result_to_entry(r) for r in all_results],
                },
                indent=2,
            )
        )
        return

    if not all_results:
        print("No benchmarks were run.")
        return

    print(f"MaxSim scoring: Q({height}x{depth}) vs D({width}x{depth})")
    print()
    table_rows = [
        {
            "Library": result.library,
            "Precision": result.display_signature,
            "GSO/s": f"{result.gso_per_sec:.2f}",
            "Time": format_duration(result.duration_secs),
        }
        for result in all_results
    ]
    print_results_table(table_rows)


if __name__ == "__main__":
    main()
